### ISSUE TITLE
Testing for x-initrd.mount in should_umount() broken?

### ISSUE THREAD

**Issue by 3chas3** [systemd/systemd#4817]
### Submission type

  - [x] Bug report
  - [ ] Request for enhancement (RFE)

*NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!*

### systemd version the issue has been seen with

215-17

*NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!*

### Used distribution

Debian 8

### In case of bug report: Expected behaviour you didn't see

Marking a mount point with x-initrd.mount didn't keep systemd from attempting to unmount the filesystem during shutdown.

### In case of bug report: Unexpected behaviour you saw

Systemd compalins about not being until to unmount said filesystem.

### In case of bug report: Steps to reproduce the problem

The following routine attempts to check for x-initrd.mount in the mount options:

static bool should_umount(Mount *m) {
        MountParameters *p;

        if (path_equal(m->where, "/") ||
            path_equal(m->where, "/usr"))
                return false;

        p = get_mount_parameters(m);
        if (p && mount_test_option(p->options, "x-initrd.mount") &&
            !in_initrd())
                return false;

        return true;
}

However, p->options is derived from /proc/self/mountinfo -- the x-* flags are not sent to the kernel and therefore they will NOT appear in /proc/self/mountinfo.

**Comment by poettering** [comment]
well, nope, we get them from libmount, and that's supposed to store userspace options in "utab", and thus not lose them... Except that libmount actually drops "x-" options right now, which is a big problem for us... ping @karelzak.

**Comment by evverx** [comment]
Based on https://github.com/systemd/systemd/pull/4515#issuecomment-266420965
```
bash4.3# mount --version
mount from util-linux 2.29.102-3f7429 (libmount 2.29.102: btrfs, assert, debug)

bash4.3# mkdir -p /test
bash4.3# mount -t tmpfs -o x-initrd.mount tmpfs /test/
bash4.3# cat /run/mount/utab
SRC=tmpfs TARGET=/test ROOT=/ OPTS=x-initrd.mount
bash4.3# systemctl show --property=Conflicts test.mount
Conflicts=umount.target
```
So, there is a bug. And simple update will not fix the issue.

**Comment by 3chas3** [comment]
Odd.  Older versions of libmount (util-linux) don't seem to handling this correctly which is complicating issue.

```
[root@tma1 ciwillia]# mount --version
mount from util-linux 2.28.2 (libmount 2.28.2: selinux, btrfs, force-mountinfo, assert, debug)
[root@tma1 ciwillia]# mkdir -p /test
[root@tma1 ciwillia]# mount -t tmpfs -o x-initrd.mount tmpfs /test
[root@tma1 ciwillia]# cat /run/mount/utab
[root@tma1 ciwillia]#
```

I guess this a new behavior as of commit 0a14cc8bcc151e03c8979ed1ff4a3bee9672365b

**Comment by evverx** [comment]
Actually, we created another issue: https://github.com/systemd/systemd/issues/4902
Fix is waiting in https://github.com/systemd/systemd/pull/4965

**Comment by fbuihuu** [comment]
> Fix is waiting in #4965

I messed up this one, the fix is now waiting in #5087.

**Comment by poettering** [comment]
Let's close this issue here, and track the dependency update in #6795.
